modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/provider/CrawlerUserAgentsProviderImpl.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.redirect.internal.provider;
+
+import com.liferay.portal.configuration.metatype.bnd.util.ConfigurableUtil;
+import com.liferay.portal.kernel.util.SetUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.redirect.configuration.CrawlerUserAgentsConfiguration;
+import com.liferay.redirect.provider.CrawlerUserAgentsProvider;
+
+import java.util.Dictionary;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.osgi.service.cm.ConfigurationException;
+import org.osgi.service.cm.ManagedService;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Alicia García
+ */
+@Component(
+	configurationPid = "com.liferay.redirect.configuration.CrawlerUserAgentsConfiguration",
+	service = {CrawlerUserAgentsProvider.class, ManagedService.class}
+)
+public class CrawlerUserAgentsProviderImpl
+	implements CrawlerUserAgentsProvider, ManagedService {
+
+	@Override
+	public boolean isCrawlerUserAgent(String userAgent) {
+		if (Validator.isNull(userAgent) ||
+			SetUtil.isEmpty(_crawlerUserAgents)) {
+
+			return false;
+		}
+
+		userAgent = StringUtil.toLowerCase(userAgent);
+
+		if (_crawlerUserAgents.contains(userAgent)) {
+			return true;
+		}
+
+		for (String crawlerUserAgent : _crawlerUserAgents) {
+			if (userAgent.contains(crawlerUserAgent)) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	@Override
+	public void updated(Dictionary<String, ?> dictionary)
+		throws ConfigurationException {
+
+		_crawlerUserAgents = _getCrawlerUserAgents();
```

Comment:
This looks strange. Shouldn't you update the `_crawlerUserAgents` based on the dictionary contents? Also, having the updated event hook doesn't make the activate method redundant?

Reply:
> Also, having the updated event hook doesn't make the activate method redundant?

This is redundant, but the `updated` it's only called when the configuration is **saved** so if you let the configuration as default and you do not touch it, the `_crawlerUserAgents` is never initialized so is `null`, so the activate, in my opinion is needed.

So, the `updated` is needed too (by the `managedService`)

@adolfopa 
What do you think of removing the managedService and add the `@Modified `on the `activate` method?

Reply:
Sounds good.